---
title: "Beta diversity variability - function"
author: "Abby"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document
---

```{r setup, include=FALSE, message = FALSE}

require(rmarkdown)
require(knitr)
require(RColorBrewer) 
require(cowplot)
require(tidyverse)
require(stringr)
require(dplyr)
require(ape)
require(vegan)

opts_knit$set(root.dir = "/Users/abby/Documents/Projects/dietstudy/")
opts_chunk$set(echo = TRUE, fig.path = "Figs/", dev = c("png", "pdf"), dpi = 300)

```

```{r load long map, echo = FALSE}
map <- read.table("data/maps/SampleID_map.txt", sep = "\t", header = T, comment = "")
map_username <- read.table("data/maps/UserName_map.txt", sep = "\t", header = T, comment = "")

# set up colors
cols <- brewer.pal(12, "Set3")
cols2 <- colorRampPalette(cols)
```


```{r import beta div, include =F, message = F}
# import beta diversity
food_beta <- read.delim("data/processed_food/dhydrt_beta/unweighted_unifrac_dhydrt.otu.txt", sep = "\t", row.names = 1)


# function
tax <- read.delim("data/shogun_results_cleaned/shogun_function/filtered_functions/embalmer_taxatable_clean.species.kegg.modules.txt", sep ="\t", row.names = 1)
# drop the silly extra space
rownames(tax) <- gsub(" ", "", rownames(tax))

# drop rare keggs
tax <- tax[rowMeans(tax) > 1,]
tax_beta <- as.matrix(vegdist(t(tax), method = "euclidian"))

# import matching maps
food_map <- read.table("data/maps/food_map.txt", header = T, sep = "\t", comment = "")
tax_map <- read.table("data/maps/taxonomy_norm_map.txt", header = T, sep = "\t", comment = "")

# calculate mean of the mean distance to self for each sample, for each UserName (food)
mydist <- NULL

for (i in unique(food_map$UserName)) {
  name <- food_map[food_map$UserName == i,1]
  name <- as.vector(name)
  tmp <- food_beta[rownames(food_beta) %in% name, colnames(food_beta) %in% name]
  num <- rowMeans(tmp, na.rm = T)
  num <- cbind(i,num)
  mydist <- rbind(num, mydist)
  colnames(mydist) <- c("UserName", "food_mean_beta_div")
}

food_dist <- as.data.frame(mydist)
food_dist$food_mean_beta_div <- as.character(food_dist$food_mean_beta_div)
food_dist$food_mean_beta_div <- as.numeric(food_dist$food_mean_beta_div)


mydist <- NULL

for (i in unique(tax_map$UserName)) {
  name <- tax_map[tax_map$UserName == i,1]
  name <- as.vector(name)
  tmp <- tax_beta[rownames(tax_beta) %in% name, colnames(tax_beta) %in% name]
  num <- rowMeans(tmp, na.rm = T)
  num <- cbind(i,num)
  mydist <- rbind(num, mydist)
  colnames(mydist) <- c("UserName", "mean_beta_div")
}

tax_dist <- as.data.frame(mydist)
tax_dist <- droplevels(tax_dist)
tax_dist$mean_beta_div <- as.character(tax_dist$mean_beta_div)
tax_dist$mean_beta_div <- as.numeric(tax_dist$mean_beta_div)
```


```{r food div per subject, echo = F}
# Get average microbiome distance to self 
# This will also be used to order the plot
micro_order <- tax_dist %>% 
  group_by(UserName) %>% 
  dplyr::summarise(ave_cloud = mean(mean_beta_div, na.rm=T))
# sort by mean
micro_order <- micro_order[order(micro_order$ave_cloud),]
# get factor names for ordering
ord_factor <- as.character(micro_order$UserName)

# calculate median food distance to self
food_order <- food_dist %>% 
  group_by(UserName) %>% 
  dplyr::summarize(food_cloud = mean(food_mean_beta_div))

```


```{r stools v cloud size, echo = F, fig.height=4, fig.width=4}

stool_n <- map %>% 
  dplyr::group_by(UserName) %>% 
  dplyr::count(fecal.status)

# drop the NAs
stool_n <- stool_n %>% 
  filter(!is.na(fecal.status), !str_detect(UserName,"Blank"))

# make wide
stool_n <- spread(stool_n, fecal.status, n, fill = NA)

stool_n$percentmissed <- ifelse(is.na(stool_n$`0`), 0, (stool_n$`0`/stool_n$`1`)*100)

#subset to username that we have in the beta_div cloud size varaible
stool_n <- stool_n[stool_n$UserName %in% micro_order$UserName,]

stool_n <- merge(stool_n, micro_order)

new_cols <- (sample(cols2(34)))

ggplot(stool_n, aes(percentmissed, ave_cloud)) +
  geom_smooth(method = "lm", color = "black", fill = "grey") + 
  geom_point(aes(fill = UserName), size = 3, color = new_cols) +
  geom_point(aes(fill = UserName), shape = 1, size =3, color = "black") +
  theme(legend.position = "none",
        axis.title = element_text(size = 10)) +
  labs(x = "Stool frequency (NBMD/BMD %)", y = "Mean distance to self")

# no correlation
cor.test(sqrt(stool_n$percentmissed), stool_n$ave_cloud, method = "spearman", exact = F)

# clearly there are two outliers and those poeple have far less frequent stooling
```


```{r food plot, echo = F}

# set colors
set.seed(7)
subject_cols <- (sample(cols2(34)))

plot_food_per_subject <- food_dist # %>%
  #filter(UserName != "MCTs06", #Infrequent bowel movements
       #  UserName != "MCTs29"  #Infrequent bowel movements
       #  ) 
plot_food_per_subject <- merge(plot_food_per_subject, map_username)

# reorder by ordered factor
plot_food_per_subject$UserName <- factor(plot_food_per_subject$UserName, levels = ord_factor)

# make plot (switch directionality)
diet <- ggplot(data = plot_food_per_subject, aes(x = UserName, y = food_mean_beta_div, fill = UserName)) +
  geom_boxplot(fill = subject_cols, outlier.shape = NA, lwd = .25) +
  geom_jitter(shape = 21, fill = NA, size = 0.5, width = 0.2, alpha = 0.5) +
  theme_classic() +
  theme(axis.title.x = element_text(size = 7), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "none") +
  labs(y = "Diet beta diversity variability (Unweighted Unifrac)", x = " ") +
  coord_flip()


```

```{r microbiome div per subject, echo = F}

plot_tax_per_subject <- tax_dist #%>%
  #filter(UserName != "MCTs06", #Infrequent Bowel Movements
        # UserName != "MCTs29"  #Infrequent Bowel Movements
        # ) 
plot_tax_per_subject <- merge(plot_tax_per_subject, map_username)

plot_tax_per_subject$UserName <- gsub("MCTs", "", plot_tax_per_subject$UserName)
ord_factor <- gsub("MCTs", "", ord_factor)

# apply ordering
plot_tax_per_subject$UserName <- factor(plot_tax_per_subject$UserName, levels = ord_factor)

# plot taxonomy diversity
micro_tax <- ggplot(data = plot_tax_per_subject, aes(x = UserName, y = mean_beta_div, fill = UserName)) +
  geom_boxplot(fill = subject_cols, outlier.shape = NA, lwd = .25) +
  geom_jitter(shape = 21, fill = NA, size = 0.5, width = 0.2, alpha = 0.5) +
  theme_classic() +
  theme(axis.title.x = element_text(size = 7),
        axis.title.y = element_text(size = 7),
        axis.text.y = element_text(size = 4),
        legend.position = "none") +
  labs(y = "Microbiome beta diversity variabitliy (CLR Euclidean)", x = "") +
  coord_flip()
  
```

```{r plotdietmicrobiome, echo = F, message = F, fig.width=11, fig.height=3}
plot_grid(micro_tax, diet, #labels = c("D", "E") , 
          ncol = 2, align = "h", rel_widths = c(1.0,1.0))

```


```{r micro v food cloud size, echo = F, fig.height=4, fig.width=4}
plot <- merge(micro_order, food_order)

plot <- plot %>%
  filter(#UserName != "MCTs11", #Soylent
         #UserName != "MCTs12", #Soylent
         #UserName != "MCTs02", #Dropped
         #UserName != "MCTs17", #Dropped
         #UserName != "MCTs30", #Dropped
         #UserName != "MCTs06", #Infrequent bowel movements
         #UserName != "MCTs29"  #Infrequent bowel movements
         ) 

ggplot(plot, aes(y = food_cloud, x = ave_cloud)) +
  geom_smooth(method = "lm", color = "black", fill = "grey") + 
  geom_point(aes(fill = UserName), size = 3, color = subject_cols) +
  geom_point(aes(fill = UserName), shape = 1, size =3, color = "black") +
  theme(legend.position = "none",
        axis.title = element_text(size = 10)) +
  labs(y = "Food mean distance to self", x = "Microbiome mean distance to self")

cor.test(plot$food_cloud, plot$ave_cloud, method = "spearman")


```



Fold difference for the variation in non-outlier microbiome cloud size:
```{r fold difference, echo = F}
# calculate the fold difference for the mean variation
# first drop subjects
micro_order <- micro_order #%>% filter(UserName != "MCTs06", UserName != "MCTs29")

max(micro_order$ave_cloud)/min(micro_order$ave_cloud)

food_order <- food_order #%>% filter(UserName != "MCTs06", UserName != "MCTs29")

max(food_order$food_cloud)/min(food_order$food_cloud)
```



```{r microbes correlate with beta diversity, echo = F}

# # first get tax average per subject starting from raw counts again
# tax <- read.delim("data/processed_tax/taxonomy_norm.biom", row = 1)
# 
# # Drop low reads
# tax <- tax[, colSums(tax) >=20000]
# 
# # Drop blank with high reads
# tax <- tax[, colnames(tax) != "Blank.D02."]
# 
# # Summarizing at different levels - makes changes to everything downstream
# split <- strsplit(rownames(tax),";")             # Split and rejoin on lv7 to get species level
# taxaStrings <- sapply(split,function(x) paste(x[1:2],collapse=";"))
# tax <- rowsum(tax,taxaStrings)              # Collapse by taxonomy name

#initiate the dataframe with the correct information
tax_subject <- data.frame(matrix(nrow=nrow(tax), ncol=0))
rownames(tax_subject) <- rownames(tax)

#loop through and summarize by username
for (i in unique(tax_map$UserName)){
  subgroup <- tax_map[tax_map$UserName == i,]
  if(dim(subgroup)[1] <= 1){ 
    NULL
  }else{
  sub <- tax[(colnames(tax) %in% subgroup[,"X.SampleID"])]
  tmp <- as.data.frame(rowMeans(sub))
  colnames(tmp) <- i
  tax_subject <- cbind(tax_subject,tmp)
  } 
}

# Normalize
#tax_subject <- sweep(tax_subject, 2, colSums(tax_subject), "/")

# Sort by average abundance
tax_subject <- tax_subject[order(rowMeans(tax_subject),decreasing=T),]

# Drop low abundance taxa
#tax_subject <- tax_subject[rowMeans(tax_subject) >= 0.001,]

# drop outliers
#tax_subject <- tax_subject %>% select(-c(MCTs06,MCTs29))


# subset to same number of subjects
tax_subject <- tax_subject[,colnames(tax_subject) %in% micro_order$UserName]
micro_order <- micro_order[micro_order$UserName %in% colnames(tax_subject),]
micro_order <- micro_order[order(micro_order$UserName),]

# get correlations pvalues
pvals <- apply(tax_subject, 1, function(x) {cor.test(x, micro_order$ave_cloud, method = "spearman")$p.value})
cors <- apply(tax_subject, 1, function(x) {cor(x, micro_order$ave_cloud, method = "spearman")})

qvals <- p.adjust(pvals, method = "fdr")
qvals <- qvals[qvals <= 0.32]

qvals <- as.data.frame(qvals)
cors <- as.data.frame(cors)

# nothing correlated significantly with median beta diversity if we leave out the outliers
# if we include the outliers, then
out <- merge(qvals, cors, by = 0)

out <- out[order(out$qvals, decreasing = F),]
out <- remove_rownames(out)

colnames(out)[1] <- "Taxa"

out

out <- out[order(out$cors, decreasing = T),]

out <- out[1:10,]

```


# plot significant tax versus median beta diversity

```{r sig plots, echo = F, fig.height=4, fig.width=4}
plot <- as.data.frame(t(tax_subject))
plot <- plot[,colnames(plot) %in% out$Taxa]
plot <- rownames_to_column(plot, var = "UserName")
plot <- merge(micro_order, plot)

colnames(plot) <- gsub("?.*s__", "", colnames(plot))
colnames(plot) <- gsub("\\/", "_", colnames(plot))
colnames(plot) <- gsub("\\-", "_", colnames(plot))

out$Taxa <- gsub("?.*s__", "", out$Taxa)
out$Taxa <- gsub("\\/", "_", out$Taxa)
out$Taxa <- gsub("\\-", "_", out$Taxa)

out <- remove_rownames(out)
out <- column_to_rownames(out, var = "Taxa")


for (i in colnames(plot[,3:ncol(plot)])) {
  rho <- paste("italic(r) == ", round(out[i,"cors"], 3))
  q <- paste("italic(qval) ==", round(out[i,"qvals"], 3))
  labels = data.frame(x = quantile(plot[,i])[3], y =quantile(plot[,"ave_cloud"])[1], 
                       label = paste(rho, q, sep = " ~ "))
   
   
   print(ggplot(data = plot, aes_string(x = i, y = "ave_cloud")) + 
           geom_smooth(method = "lm", color = "black", fill = "grey") +
           geom_point(aes(fill = UserName), color = subject_cols, size = 3) +
           geom_point(aes(fill = UserName), shape = 1, size =3, color = "black") +
           theme(legend.position = "none") +
           geom_text(data = labels, aes(x=x, y=y,
            label = label),
            parse = T, size = 3) +
           labs(y = "Microbiome Function Variability")) 
}


```

```{r foods beta cloud, echo=F}
# Are any foods associated with beta-diversity cloud size?
food <- read.table("data/processed_food/dhydrt.otu.txt", header = T, sep = "\t", comment = "")

# get the foods in the correct format
food <- remove_rownames(food)
food <- column_to_rownames(food, var = "taxonomy")
food <- food %>% select(-X.FoodID)

# collapse by taxa level 3
split <- strsplit(rownames(food),";")             # Split and rejoin on lv7 to get species level
foodStrings <- sapply(split,function(x) paste(x[1:3],collapse=";"))
food <- rowsum(food,foodStrings)              # Collapse by taxonomy name

# Sort by average abundance
food <- food[order(rowMeans(food),decreasing=T),]

# summarize by person
#initiate the dataframe with the correct information
food_subject <- data.frame(matrix(nrow=nrow(food), ncol=0))
rownames(food_subject) <- rownames(food)

# fix naming of food_map
colnames(food_map)[1] <- "X.SampleID"

#loop through and summarize by username
for (i in unique(food_map$UserName)){
  subgroup <- food_map[food_map$UserName == i,]
  if(dim(subgroup)[1] <= 4){ 
    NULL
  }else{
  sub <- food[(colnames(food) %in% subgroup[,"X.SampleID"])]
  tmp <- as.data.frame(rowMeans(sub))
  colnames(tmp) <- i
  food_subject <- cbind(food_subject,tmp)
  } 
}

# Normalize
food_subject <- sweep(food_subject, 2, colSums(food_subject), "/")

# Sort by average abundance
food_subject <- food_subject[order(rowMeans(food_subject),decreasing=T),]

# drop low abundance foods
food_subject <- food_subject[rowMeans(food_subject) >= 0.001,]

#food_subject <- food_subject %>% select(-MCTs09)


# subset to same number of subjects
food_subject <- food_subject[,colnames(food_subject) %in% micro_order$UserName]
micro_order <- micro_order[micro_order$UserName %in% colnames(food_subject),]
micro_order <- micro_order[order(micro_order$UserName),]

# get correlations pvalues
pvals <- apply(food_subject, 1, function(x) {cor.test(x, micro_order$ave_cloud, method = "spearman", exact = F)$p.value})
cors <- apply(food_subject, 1, function(x) {cor(x, micro_order$ave_cloud, method = "spearman")})

qvals <- p.adjust(pvals, method = "fdr")
qvals <- qvals[qvals <= 0.25]

qvals <- as.data.frame(qvals)
cors <- as.data.frame(cors)

# nothing correlated significantly with median beta diversity if we leave out the outliers
# if we include the outliers, then
out <- merge(qvals, cors, by = 0)

out <- out[order(out$qvals, decreasing = F),]
out <- remove_rownames(out)

colnames(out)[1] <- "Species"

out
```

```{r food sig plots, echo = F, fig.height=4, fig.width=4}
plot <- as.data.frame(t(food_subject))
plot <- plot[,colnames(plot) %in% out$Species]
plot <- rownames_to_column(plot, var = "UserName")
plot <- merge(micro_order, plot)

colnames(plot) <- gsub("?.*L3_", "", colnames(plot))
colnames(plot) <- gsub("\\/", "_", colnames(plot))
colnames(plot) <- gsub("\\-", "_", colnames(plot))

out$Species <- gsub("?.*L3_", "", out$Species)
out$Species <- gsub("\\/", "_", out$Species)
out$Species <- gsub("\\-", "_", out$Species)

out <- remove_rownames(out)
out <- column_to_rownames(out, var = "Species")

for (i in colnames(plot[,3:ncol(plot)])) {
  rho <- paste("italic(r) == ", round(out[i,"cors"], 3))
  q <- paste("italic(qval) ==", round(out[i,"qvals"], 3))
  labels = data.frame(x = quantile(plot[,i])[3], y =quantile(plot[,"ave_cloud"])[1], 
                       label = paste(rho, q, sep = " ~ "))
   
   
   print(ggplot(data = plot, aes_string(x = i, y = "ave_cloud")) + 
           geom_smooth(method = "lm", color = "black", fill = "grey") +
           geom_point(aes(fill = UserName), color = subject_cols[1:34], size = 3) +
           geom_point(aes(fill = UserName), shape = 1, size =3, color = "black") +
           theme(legend.position = "none") +
           geom_text(data = labels, aes(x=x, y=y,
            label = label),
            parse = T, size = 3) +
           labs(y = "Microbiome mean distance to self")) 
}


```


```{r food pcoa, echo = F, fig.width=7.5, fig.height = 4}

#betad as matrix 
betadiv <- data.frame(pcoa(food_beta)$vectors)

# move rownames to a column
betadiv <- rownames_to_column(betadiv, var = "X.SampleID")

#merge map and PCOA by SampleID - select both lines and run together - won't work otherwise
betadiv <- inner_join(betadiv, map, by = 'X.SampleID')

# drop dropouts
betadiv <- betadiv %>%
  filter(#UserName != "MCTs11", #Soylent
         #UserName != "MCTs12", #Soylent
         UserName != "MCTs02", #Dropped
         UserName != "MCTs17", #Dropped
         UserName != "MCTs30" #Dropped
         #UserName != "MCTs06", #Infrequent bowel movements
         #UserName != "MCTs29"  #Infrequent bowel movements
         ) 

Axis.1.2 <- ggplot(betadiv, aes(x = Axis.1, y = Axis.2, fill = UserName)) +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 2) +
  stat_ellipse(aes(color = UserName), type = "norm", linetype = 2, level = 0.95, alpha = 0.3) +
  scale_fill_manual(values = subject_cols) + 
  scale_color_manual(values = subject_cols) +
  labs(title="Dietary Beta-diversity \n(PD whole tree)") +
  theme_classic() +                                                                      
  theme(plot.title = element_text(size = 14, hjust = 0.5),
        legend.position = "none")


Axis.2.3 <- ggplot(betadiv, aes(x = Axis.3, y = Axis.2, fill = UserName)) +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 2) +
  stat_ellipse(aes(color = UserName), type = "norm", linetype = 2, level = 0.95, alpha = 0.3) +
  scale_fill_manual(values = subject_cols) + 
  scale_color_manual(values = subject_cols) +
  labs(title="Dietary Beta-diversity \n(PD whole tree)") +
  theme_classic() +                                                                      
  theme(plot.title = element_text(size = 14, hjust = 0.5),
        legend.position = "none")

plot_grid(Axis.1.2, Axis.2.3)

```

```{r food by day, echo = F, fig.width=10.5, fig.height=24}
Axis1.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.1, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 3),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))
  

Axis2.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.2, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 3),
        axis.ticks.x = element_blank(),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))


Axis3.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.3, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 3),
        axis.ticks.x = element_blank(),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))
  

plot_grid(Axis1.day,Axis2.day, Axis3.day, ncol = 1)
```

Next up, try coloring this by prevatella/bacteroides etc. 
```{r taxa pcoa, echo = F, fig.width=7.5, fig.height = 4}

#betad as matrix 
betadiv <- data.frame(pcoa(tax_beta)$vectors)

# move rownames to a column
betadiv <- rownames_to_column(betadiv, var = "X.SampleID")

#merge map and PCOA by SampleID - select both lines and run together - won't work otherwise
betadiv <- inner_join(betadiv, map, by = 'X.SampleID')

# drop dropouts
betadiv <- betadiv %>%
  filter(#UserName != "MCTs11", #Soylent
         #UserName != "MCTs12", #Soylent
         UserName != "MCTs02", #Dropped
         UserName != "MCTs17", #Dropped
         UserName != "MCTs30" #Dropped
         #UserName != "MCTs06", #Infrequent bowel movements
         #UserName != "MCTs29"  #Infrequent bowel movements
         ) 

Axis.1.2 <- ggplot(betadiv, aes(x = Axis.1, y = Axis.2, fill = UserName)) +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 2) +
  stat_ellipse(aes(color = UserName), type = "norm", linetype = 2, level = 0.95, alpha = 0.3) +
  scale_fill_manual(values = subject_cols) + 
  scale_color_manual(values = subject_cols) +
  labs(title="Microbiome Beta-diversity \n(Euclidean - CLR)") +
  theme_classic() +                                                                      
  theme(plot.title = element_text(size = 14, hjust = 0.5),
        legend.position = "none")


Axis.2.3 <- ggplot(betadiv, aes(x = Axis.3, y = Axis.2, fill = UserName)) +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 2) +
  stat_ellipse(aes(color = UserName), type = "norm", linetype = 2, level = 0.95, alpha = 0.3) +
  scale_fill_manual(values = subject_cols) + 
  scale_color_manual(values = subject_cols) +
  labs(title="Microbiome Beta-diversity \n(Euclidean - CLR)") +
  theme_classic() +                                                                      
  theme(plot.title = element_text(size = 14, hjust = 0.5),
        legend.position = "none")


plot_grid(Axis.1.2, Axis.2.3)


```

```{r tax by day, echo = F, fig.width=10.5, fig.height=24}
Axis1.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.1, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 3),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))
  

Axis2.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.2, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 3),
        axis.ticks.x = element_blank(),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))


Axis3.day <- ggplot(betadiv, aes(x = StudyDayNo, y = Axis.3, fill = UserName, droplevels = TRUE)) +
  geom_line(color = "grey") +
  geom_point(color = "black", alpha=0.75, pch = 21, size = 3) +
  scale_fill_manual(values = subject_cols) +
  facet_wrap(~UserName)+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 3),
        axis.ticks.x = element_blank(),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size = 7),
        strip.background = element_rect(color = "grey", fill = "white"))
  

plot_grid(Axis1.day,Axis2.day, Axis3.day, ncol = 1)

```

Is dietary beta diversity associated with the ability to sucessfully match microbiome and diet with procrutes?

```{r diet beta v procrutes sig, echo = F, fig.height = 4, fig.width=4}
# set colors
procrustes <- map_username[map_username$UserName %in% food_order$UserName,]
procrustes <- merge(procrustes, food_order)
procrustes <- merge(procrustes, micro_order)

# drop soylent
procrustes <- procrustes %>% filter(!(UserName %in% c("MCTs11", "MCTs12")))

subject_cols <- subject_cols[1:32]

# food cloud size affects ability to predict microbiome from foods                 
t.test(procrustes$food_cloud~as.factor(procrustes$Procrustes))
cor.test(procrustes$food_cloud,procrustes$Monte.Carlo.p.value)

ggplot(data = procrustes, aes(x = Monte.Carlo.p.value, y = food_cloud)) +
  geom_smooth(method = "lm", color = "black", fill = "grey") +
  geom_point(aes(fill = UserName), shape = 21, size = 3, alpha = 1) +
  scale_fill_manual(values = c(subject_cols)) +
  theme_classic() +
  theme(legend.position = "none") +
  labs(x = "Monte Carlo P-value", y = "Dietary variability")


# Microbiome alpha diveristy does not affect ability to predict microbiome from diet
t.test(procrustes$ave_cloud ~as.factor(procrustes$Procrustes))
cor.test(procrustes$ave_cloud,procrustes$Monte.Carlo.p.value, method = "spearman", exact = F)


ggplot(data = procrustes, aes(x = Procrustes, y = ave_cloud)) +
  geom_boxplot(aes(fill = Procrustes), outlier.shape = NA, lwd = .25)+
  geom_jitter(aes(fill = UserName), shape = 21, size = 3, width = 0.25, alpha = 1) +
  scale_fill_manual(values = c(subject_cols, "lightgrey", "darkgrey")) +
  theme_classic() +
  theme(legend.position = "none") +
  labs(x = "Procrustes result", y = "Microbiome variability")

# So the ratio between micorbiome variability and diet variabitily might be whats improtant
procrustes$m.f <- procrustes$ave_cloud/procrustes$food_cloud
t.test(procrustes$m.f ~ as.factor(procrustes$Monte.Carlo.p.value < 0.05))

new_cols <- c(subject_cols, "lightgrey", "darkgrey")

ggplot(data = procrustes, aes(x = as.factor(procrustes$Monte.Carlo.p.value < 0.05), y = m.f)) +
  geom_boxplot(outlier.shape = NA, lwd = .25, fill = "lightgrey") +
  geom_jitter(aes(fill = UserName), shape = 21, size = 3, width = 0.25, alpha = 1) +
  scale_fill_manual(values = subject_cols) +
  theme_classic() +
  theme(legend.position = "none") +
  labs(x = "Procrustes result", y = "Microbiome variability/diet variability")

  
  
```
